doc/compiler/optimizer/BenefitInliner.md

```diff
@@ -0,0 +1,94 @@
+``` 
+
+Note: `generateControlFlowGraph()` and `performAbstractInterpretation()` need language-specific implementation.
+
+The responsibilities of the **abstract interpretation**:
+> 1. Identifying call sites. Call sites will be added to the **IDT** as **IDTNodes**
+
+> 2. Generating inlining method summaries. (Refer to **InliningMethodSummary.cpp** for details )
+
+> 3. Simulating the program state. Currently, the useful part of the program state is the arguments ( represented by **AbsArguments**) passed to each call site. Combining those arguments with the inlining method summary, we can get the static benefit of inlining this method.
+
+### 4. AbsVisitor.hpp & IDTBuilderVisitor
+* **AbsVisitor** is an interface for the derived classes to define customized callback methods during abstract interpretation. Currently, there is only one method defined in **AbsVisitor** called `AbsVistor::visitCallSite`. **IDTBuilderVisitor** inherits from **AbsVisitor** and defines it own implementation of `visitCallSite`. Whenever we encounter a call site during abstract interpretation, `AbsVistor::visitCallSite` will be invoked. If **AbsVisitor** is an **IDTBuilderVisitor**, `IDTBuilder::addChild` will be called in order to add an **IDTNode** to the **IDT**.
+
+### 5. InliningMethodSummary.cpp
+
+**PotentialOptimization** is the base class for all the potential optimizations such as branch folding and null-check folding, etc. It takes **VPConstraint** as a dependency to serve as the constraint (lattice). Those constraints specify the maximal safe values to make the optimizations happen. `PotentialOptimization::predicate` tests against an argument value to tell whether or not this abstract argument is a safe value to unlock the optimization.
+
+**InliningMethodSummary** captures potential optimization opportunities of inlining one particular method. `InliningMethodSummary::predicates` tells the total optimizations that will be unlocked given the abstract argument value.
+
+### 6. AbsValue.cpp, AbsStackMachineState.cpp & AbsState.cpp
+
+**AbsValue** is the abstract representation of value. It uses **VPConstraint** as the lattice to represent the value. A **top** **AbsValue** is the least accurate abstract value, analogous to the unknown value. A **top** **AbsValue** simply has constraint being NULL. `AbsValue::merge` relies on `VPConstraint::merge` for merging various type of values. **AbsValue** comes with a set of factory methods for creating different kinds of abstract values. 
+
+* **AbsState**
+    * **AbsStackMachineState**
+
+**AbsState** is the abstract representation of the program state. **AbsStackMachineState** inherits from **AbsState** and it is used for stack machine language. It takes **AbsLocalVarArray** and **AbsOpStack** as dependencies. **AbsState** is used during abstract interpretation. `TR::Block` takes **AbsState** as an dependency. 
```

Comment:
Why does `TR::Block` need to know about `AbsState`?

Reply:
So the abstract interpretater walks the blocks of the CFG in reverse-post order, interpreting the blocks one by one and merging the absStates of the block's predecessors if it has multiple predecessors by the control flow.  Let TR::Block has an absState is useful.

Reply:
I definitely don't think `TR::Block` should be modified for this work. The abstract interpreter state is only meaningful while doing abstract interpretation and isn't a general property of a block, so I'd much prefer keeping a separate mapping between blocks and states.

Reply:
I agree with @aviansie-ben and @Leonardo2718 on this one.

Reply:
`AbsState` removed